pkg/provider/azure_loadbalancer.go

```diff
@@ -2851,36 +2851,50 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
 		}
 	}
 
-	sourceRanges, err := servicehelpers.GetLoadBalancerSourceRanges(service)
+	accessControl, err := loadbalancer.NewAccessControl(service)
 	if err != nil {
+		klog.ErrorS(err, "Failed to parse access control configuration for service", "service", service.Name)
 		return nil, err
 	}
-	serviceTags := getServiceTags(service)
-	if len(serviceTags) != 0 {
-		delete(sourceRanges, consts.DefaultLoadBalancerSourceRanges)
-	}
 
-	sourceAddressPrefixes := map[bool][]string{}
-	if (sourceRanges == nil || servicehelpers.IsAllowAll(sourceRanges)) && len(serviceTags) == 0 {
-		if !requiresInternalLoadBalancer(service) || len(service.Spec.LoadBalancerSourceRanges) > 0 {
-			sourceAddressPrefixes[false] = []string{"Internet"}
-			sourceAddressPrefixes[true] = []string{"Internet"}
-		}
-	} else {
-		for _, ip := range sourceRanges {
-			if ip == nil {
-				continue
-			}
-			isIPv6 := net.ParseIP(ip.IP.String()).To4() == nil
-			sourceAddressPrefixes[isIPv6] = append(sourceAddressPrefixes[isIPv6], ip.String())
+	var (
+		sourceRanges          = accessControl.SourceRanges()
+		allowedServiceTags    = accessControl.AllowedServiceTags()
+		allowedIPRanges       = accessControl.AllowedIPRanges()
+		sourceAddressPrefixes = map[bool][]string{
+			false: accessControl.IPV4Sources(),
+			true:  accessControl.IPV6Sources(),
 		}
-		sourceAddressPrefixes[false] = append(sourceAddressPrefixes[false], serviceTags...)
-		sourceAddressPrefixes[true] = append(sourceAddressPrefixes[true], serviceTags...)
+	)
+
+	if len(sourceRanges) != 0 && len(allowedIPRanges) != 0 {
+		// Block the service and return error if both of spec.loadBalancerSourceRanges and annotation are specified
+		klog.Warningf("Service %s is using both of spec.loadBalancerSourceRanges and annotation %s.", service.Name, consts.ServiceAnnotationAllowedIPRanges)
+		return nil, fmt.Errorf(
+			"both of spec.loadBalancerSourceRanges and annotation %s are specified for service %s, which is not allowed",
+			consts.ServiceAnnotationAllowedIPRanges, service.Name,
+		)
+	}
+	if len(sourceRanges) != 0 && len(allowedServiceTags) != 0 {
+		// Suggesting to use aks custom annotation instead of spec.loadBalancerSourceRanges
+		klog.Warningf(
+			"Service %s is using both of spec.loadBalancerSourceRanges and annotation %s.",
+			service.Name, consts.ServiceAnnotationAllowedServiceTags,
+		)
+		az.Event(service, v1.EventTypeWarning, "FailedReconcileSecurityGroup", fmt.Sprintf(
```

Comment:
This message could be confusing as reconcileSecurityGroup has not failed here.

Reply:
updated. I am not sure if `ConflictConfiguration` is too general.

Reply:
Would it be better to change to DeprecatedConfigurationLoadBalancerSourceRange?

Reply:
I think `LoadBalancerSourceRange` is not deprecated.
And I am ok with `ConflictConfigurationLoadBalancerSourceRange`.

Reply:
It looks like they are not conflicted, and they are just not a recommended combo.

Reply:
Before this change, using both of them at the same time wouldn’t work either. It seems to me that there is a conflict.

And this PR does NOT change any behavior of it.

Reply:
@nilo19 when you use loadBalancerSourceRanges, kube-proxy also implements iptables rules to limit traffic to those IPs, which ends up blocking the service tags traffic even if we allow it through the NSG via the rules. This is why they don't work together - everything will configure properly, but the traffic won't be allowed through because NSG allows both the IPs and the service tags but the iptables on the node only allows the IPs. Configuring it via annotation here only enforces the restriction at the NSG and leaves kube-proxy's rules wide open, and thus works as expected.

Reply:
So `loadBalancerSourceRanges` is not deprecated and still fine to use *as long as you're not also using allowed service tags*, hence the conflict warning.